### PR TITLE
Fixes #178: Improved error handling for maps and keys options

### LINKED ISSUE #178: Missing node key field results in unexplained error on relationship writes
Look at the code carefully:  notice that "name:author" is correct but reversed, it should be "author:name".

Screen Shot 2020-09-09 at 9 52 54 AM

This is a user error, but the error makes no sense.  Managed to replicate a bug post-merge that we were looking into the other day.   We need a check before writes begin that any fields mentioned/implicated in the keys for source/target must exist.  This dataframe didn't even have a "name" field so before opening any connection, this error should be detectable/reportable.

### PR DESCRIPTION
**Pull request by utnaf**
Fixes #178 

Details in the issue. 

With this changes, the following scenario:

```scala
val musicDf = Seq(
      (1, 12, "John Henry Bonham", "Drums"),
      (2, 19, "John Mayer", "Guitar"),
      (3, 32, "John Scofield", "Guitar"),
      (4, 15, "John Butler", "Guitar")
    ).toDF("id", "experience", "name", "instrument")

    musicDf.write
        .format(classOf[DataSource].getName)
        .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
        .option("relationship", "PLAYS")
        .option("relationship.source.save.mode", "Overwrite")
        .option("relationship.target.save.mode", "Overwrite")
        .option("relationship.save.strategy", "keys")
        .option("relationship.source.labels", ":Musician")
        .option("relationship.source.node.keys", "musician_name:name,another_name:name") // <-- wrong
        .option("relationship.target.labels", ":Instrument")
        .option("relationship.target.node.keys", "instrument_name:name") // <-- wrong
        .save()
```

will produce the following error:

```
Write failed due to the following errors:
 - Schema is missing instrument_name from option `relationship.target.node.keys`
 - Schema is missing musician_name, another_name from option `relationship.source.node.keys`

The option key and value might be inverted.
```

More examples in the new tests.

Minor issues fixed in the PR:
- Jar name is wrong due to a misspelled variable in pom.xml

### CODE REVIEW

**Review by moxious — APPROVED** [review]

File: src/main/scala/org/neo4j/spark/service/MappingService.scala [hunk 1]
```diff
@@ -63,8 +65,8 @@ class Neo4jWriteMappingStrategy(private val options: Neo4jOptions)
     }
   }
 
-  private def keysStrategyConsumer(): MappingBiConsumer = new MappingBiConsumer {
-    override def accept(key: String, value: AnyRef): Unit =
+  private def keysStrategyConsumer(schema: StructType): MappingBiConsumer = new MappingBiConsumer {
```
Comment (conker84) [inline src/main/scala/org/neo4j/spark/service/MappingService.scala:68]: is `(schema: StructType)` necessary?
Can we do this?
```suggestion
  private val keysStrategyConsumer = new MappingBiConsumer {
```

File: src/main/scala/org/neo4j/spark/util/Validations.scala [hunk 2]
```diff
@@ -41,6 +45,47 @@ object Validations {
...
+      schema.missingFields(neo4jOptions.relationshipMetadata.target.nodeKeys.keySet)
+    )
+
+    val optionsWithMissingFields = missingFieldsMap.filter(_._2.nonEmpty)
+
+    if (optionsWithMissingFields.nonEmpty) {
+      throw new IllegalArgumentException("Write failed due to the following errors:\n" +
+        optionsWithMissingFields.map(field => s" - Schema is missing ${field._2.mkString(", ")} from option `${field._1}`").mkString("\n") +
+        "\n\nThe option key and value might be inverted.")
+    }
```
Comment (conker84) [inline src/main/scala/org/neo4j/spark/util/Validations.scala:86]: please use the pipe operator `|` with `stripMargin`

File: src/main/scala/org/neo4j/spark/util/Validations.scala [hunk 3]
```diff
@@ -41,6 +45,47 @@ object Validations {
...
+      schema.missingFields(neo4jOptions.relationshipMetadata.source.nodeKeys.keySet)
+    )
+    missingFieldsMap.put(
+      Neo4jOptions.RELATIONSHIP_TARGET_NODE_PROPS,
+      schema.missingFields(neo4jOptions.relationshipMetadata.target.nodeProps.keySet)
+    )
+    missingFieldsMap.put(
+      Neo4jOptions.RELATIONSHIP_TARGET_NODE_KEYS,
+      schema.missingFields(neo4jOptions.relationshipMetadata.target.nodeKeys.keySet)
+    )
```
Comment (conker84) [inline src/main/scala/org/neo4j/spark/util/Validations.scala:78]: can you switch to immutable Map like this?
```scala
val missingFieldsMap = Map(Neo4jOptions.NODE_KEYS -> schema.missingFields(neo4jOptions.nodeMetadata.nodeKeys.keySet),
Neo4jOptions.NODE_PROPS -> schema.missingFields(neo4jOptions.nodeMetadata.nodeProps.keySet)
)
```

File: src/main/scala/org/neo4j/spark/util/Neo4jImplicits.scala [hunk 4]
```diff
@@ -94,4 +94,17 @@ object Neo4jImplicits {
...
+  implicit class StructTypeImplicit(structType: StructType) {
+    def getFieldsName: Seq[String] = if (structType == null) {
+      Seq.empty
+    } else {
+      structType.map(structField => structField.name)
+    }
+
+    def missingFields(fields: Set[String]): Set[String] = {
+      val structFieldsNames = structType.getFieldsName
+      fields.filter(!structFieldsNames.contains(_))
```
Comment (conker84) [inline src/main/scala/org/neo4j/spark/util/Neo4jImplicits.scala:106]:
```suggestion
      fields.filterNot(structFieldsNames.contains(_))
```

File: src/main/scala/org/neo4j/spark/util/Neo4jImplicits.scala [hunk 5]
```diff
@@ -94,4 +94,17 @@ object Neo4jImplicits {
...
   }
 
+  implicit class StructTypeImplicit(structType: StructType) {
+    def getFieldsName: Seq[String] = if (structType == null) {
+      Seq.empty
+    } else {
+      structType.map(structField => structField.name)
+    }
+
+    def missingFields(fields: Set[String]): Set[String] = {
```
Comment (conker84) [inline src/main/scala/org/neo4j/spark/util/Neo4jImplicits.scala:104]:
```suggestion
    def getMissingFields(fields: Set[String]): Set[String] = {
```

**Review by conker84 — CHANGES_REQUESTED** [review]

**Comment by conker84** [comment]
@utnaf thank you please look at my comments

**Review by conker84 — APPROVED** [review]

**Comment by conker84** [comment]
@utnaf thank you

**Pull request merged**